### PR TITLE
refactor: use `withKeyring` method

### LINKED ISSUE #24276: [keyring-controller] Mitigate run conditions with `>^16.x`
After merging these changes (Likely released in `@metamask/keyring-controller` version `16.0.0` at the time of writing):
- https://github.com/MetaMask/core/pull/4181
- https://github.com/MetaMask/core/pull/4182
- https://github.com/MetaMask/core/pull/4199
- https://github.com/MetaMask/core/pull/4192
- https://github.com/MetaMask/core/pull/4197

`@metamask/keyring-controller` should be updated to a minimum version of (at the time of writing) `^16.0.0` to bring the changes to the client.

### Update guide

The version `16.0.0` (presumably) brings many improvements, especially in terms of run conditions mitigation. There are two main breaking changes between 15.x and 16.x:
- `KeyringController`'s methods don't return the entire controller state anymore: in case that's still needed, clients can use the `controller.state` getter
- `persistAllKeyrings` (and relative `:persistAllKeyrings` action) has been deprecated, and an alternative for methods that were deprecated in previous versions has been provided now: the `withKeyring` method
  - Clients (and other controllers) can leverage the `withKeyring` method to drop usage of `getKeyringsByType`, `getKeyringForAccount`, and `persistAllKeyrings`, by passing a keyring selector and a callback that will be executed as a mutually exclusive controller operation

### PR DESCRIPTION
**Pull request by mikesposito**
## **Description**

With `@metamask/keyring-controller` v16 a new method is available which simplify safe direct keyring interactions which are not available through KeyringController.

Through `withKeyring` it's possible to lock KeyringController's main mutex and select a specific keyring to interact with while KeyringController will be unusable concurrently. `withKeyring` also takes care of persisting keyrings and updating the controller state at the end of the operation, which makes it possible to use it in substitution of `getKeyringsByType` and `persistAllKeyrings`.

Currently, most of the direct keyring interactions in the extension are made for hardware devices and snaps.

This PR is supposed to be merged after #25033 

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/25435?quickstart=1)

## **Related issues**

Fixes: #24276 

## **Manual testing steps**

Affected workflows may include:

- Lock
- Unlock
- Add account
- Remove account
- Connect hardware wallet
- All hardware wallet interactions

## **Screenshots/Recordings**

N/A

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/scripts/metamask-controller.js [hunk 1]
```diff
@@ -3964,7 +3964,12 @@ export default class MetamaskController extends EventEmitter {
         // keyring's iframe and have the setting initialized properly
         // Optimistically called to not block MetaMask login due to
         // Ledger Keyring GitHub downtime
-        this.setLedgerTransportPreference();
+        if (this.isKeyringAvailable(LedgerKeyring.type)) {
+          await this.withKeyringForDevice(
+            { name: HardwareDeviceNames.ledger },
+            async (keyring) => this.setLedgerTransportPreference(keyring),
+          );
+        }
```
Comment (mikesposito) [inline app/scripts/metamask-controller.js:3972]: I think this could also be completely removed, since there shouldn't be any Ledger keyring right after creating (or restoring) the vault.

The `this.isKeyringAvailable(LedgerKeyring.type)` condition helps not creating the keyring in case it was not present in the first place

File: app/scripts/metamask-controller.js [hunk 2]
```diff
@@ -4161,13 +4161,11 @@ export default class MetamaskController extends EventEmitter {
...
-        await this.withKeyringForDevice(
-          { name: HardwareDeviceNames.ledger },
-          async (keyring) => this.setLedgerTransportPreference(keyring),
-        );
-      }
+    if (completedOnboarding && this.isKeyringAvailable(LedgerKeyring.type)) {
+      this.withKeyringForDevice(
+        { name: HardwareDeviceNames.ledger },
+        async (keyring) => this.setLedgerTransportPreference(keyring),
+      );
```
Comment (mikesposito) [inline app/scripts/metamask-controller.js:4168]: It doesn't look nice to leave this Promise dangling, but there was this comment on top:
> // This must be set as soon as possible to communicate to the
    // keyring's iframe and have the setting initialized properly
    // Optimistically called to not block MetaMask login due to
    // Ledger Keyring GitHub downtime
    
KeyringController safeguards should work despite this promise being left dangling

File: app/scripts/metamask-controller.js [hunk 3]
```diff
@@ -4647,13 +4710,8 @@ export default class MetamaskController extends EventEmitter {
     this.custodyController.removeAccount(address);
     ///: END:ONLY_INCLUDE_IF(build-mmi)
 
-    const keyring = await this.keyringController.getKeyringForAccount(address);
     // Remove account from the keyring
     await this.keyringController.removeAccount(address);
-    const updatedKeyringAccounts = keyring ? await keyring.getAccounts() : {};
-    if (updatedKeyringAccounts?.length === 0) {
-      keyring.destroy?.();
-    }
```
Comment (mikesposito) [inline app/scripts/metamask-controller.js:4656]: KeyringController calls `keyring.destroy()` automatically when the keyring is left with no accounts

File: app/scripts/metamask-controller.js [hunk 4]
```diff
@@ -4366,14 +4416,15 @@ export default class MetamaskController extends EventEmitter {
    * @returns {Promise<boolean>}
    */
   async forgetDevice(deviceName) {
-    const keyring = await this.getKeyringForDevice(deviceName);
+    return this.withKeyringForDevice({ name: deviceName }, async (keyring) => {
+      for (const address of keyring.accounts) {
+        await this._onAccountRemoved(address);
```
Comment (mikesposito) [inline app/scripts/metamask-controller.js:4421]: By extracting side effects from `removeAccount` we can execute side effects here without ending up in a deadlock.

The account is removed from KeyringController anyway when calling `keyring.forgetDevice()`

**Comment by Gudahtt** [comment]
I have temporarily moved this back into draft to resolve conflicts

**Comment by Gudahtt** [comment]
This has been rebased onto `develop`

File: app/scripts/metamask-controller.js (hunk 1 above)
Comment (Gudahtt) [reply app/scripts/metamask-controller.js:3951]: Unfortunately we creates a new Ledger keyring for each user irrespective of whether they are Ledger users. This does seem bad, but I am unsure what the consequences would be for stopping this practice. The comment does suggest that setting this preference as soon as possible was critical, but I don't recall why.

Could we consider continuing the questionable practice of creating a Ledger keyring for all users (as we currently do), and stop doing this in a separate PR? So that we can reduce the changes of complications from this refactor.

Comment (Gudahtt) [reply app/scripts/metamask-controller.js:3951]: We might also want to drop the `await` here, given the comment above

Comment (mikesposito) [reply app/scripts/metamask-controller.js]: Good point, we can address this in a separate PR. I removed `isKeyringAvailable` from the two occurrences of this practice for Ledger (and the `isKeyringAvailable` function itself, since it is now unused).

Perhaps a good place to set the transport preference would be in `withKeyringForDevice` since it is used for all hardware keyring interactions

File: app/scripts/metamask-controller.js [hunk 5]
```diff
@@ -4177,50 +4183,73 @@ export default class MetamaskController extends EventEmitter {
...
-  async getKeyringForDevice(deviceName, hdPath = null) {
+  /**
+   * Select a hardware wallet device and execute a
+   * callback with the keyring for that device.
+   *
+   * Note that KeyringController state is not updated before
+   * the end of the callback execution, and calls to KeyringController
+   * methods within the callback can lead to deadlocks.
+   *
+   * @param {object} deviceSelector - The device selector
```
Comment (Gudahtt) [inline app/scripts/metamask-controller.js:4194]: Nit: Perhaps we can organize these parameters differently. `deviceSelector` implies that the values are used to select the keyring, but that is not true of `hdPath`. It looks like `hdPath` is set on the keyring that we find, it isn't used for selecting at all.

Maybe we can give this a more generic name like `options` so that we can specify that the `name` is used for selection, but the `hdPath` is set on whatever keyring we find. Or maybe we could remove the parameter completely, and set the `hdPath` inside the callback in the three places we use this parameter.

File: app/scripts/metamask-controller.js [hunk 6]
```diff
@@ -4395,7 +4441,7 @@ export default class MetamaskController extends EventEmitter {
     const accounts = this.accountsController.listAccounts();
 
     const { identities } = this.preferencesController.store.getState();
-    return { unlockedAccount, identities, accounts };
+    return { unlockedAccount: account.address, identities, accounts };
```
Comment (Gudahtt) [inline app/scripts/metamask-controller.js:4444]: Nit: was this change needed? I'd expect this to be equivalent:
```suggestion
    return { unlockedAccount, identities, accounts };
```

Comment (mikesposito) [reply app/scripts/metamask-controller.js:4444]: If I remember correctly, the address returned directly by the keyring can be lowercase, while the one coming from accounts controller is checksummed (or the other way around). Since we don't know how the specific keyring will return the address, I thought that it would be safer this way and would fix unit tests

File: app/scripts/metamask-controller.js [hunk 7]
```diff
@@ -4556,20 +4602,8 @@ export default class MetamaskController extends EventEmitter {
...
-    this.custodyController.removeAccount(address);
-    ///: END:ONLY_INCLUDE_IF(build-mmi)
-
-    const keyring = await this.keyringController.getKeyringForAccount(address);
-    // Remove account from the keyring
+    await this._onAccountRemoved(address);
     await this.keyringController.removeAccount(address);
-    const updatedKeyringAccounts = keyring ? await keyring.getAccounts() : {};
-    if (updatedKeyringAccounts?.length === 0) {
-      keyring.destroy?.();
```
Comment (Gudahtt) [inline app/scripts/metamask-controller.js:4571]: Are we no longer calling `destroy` on empty keyrings?

File: app/scripts/metamask-controller.js (hunk 4 above)
Comment (Gudahtt) [reply app/scripts/metamask-controller.js:4323]: Hmm. These side effects are a little strange (permissions should be automatically removed when an account is deleted, for instance). But I see that this is similar to how it already works today. Seems like a reasonable solution to me

File: app/scripts/metamask-controller.js (hunk 7 above)
Comment (Gudahtt) [reply app/scripts/metamask-controller.js:4571]: Oh I see, this is handled automatically inside the KeyringController, this was redundant

File: app/scripts/metamask-controller.js (hunk 6 above)
Comment (Gudahtt) [reply app/scripts/metamask-controller.js:4444]: Previously we were using `this.keyringController.addNewAccountForKeyring`, which returns lower-cased addresses. We lower-case the address above to normalize it.

**Review by Gudahtt — CHANGES_REQUESTED** [review]
Looks great! Just one pending issue about the `unlockedAddress` casing, and one nit.

File: app/scripts/metamask-controller.js (hunk 6 above)
Comment (mikesposito) [reply app/scripts/metamask-controller.js:4444]: Yes exactly, since that is now inside `withKeyringForDevice`' callback we can't call `KeyringController`. We can normalize it

Though, it looks like it is only used for tests

Comment (mikesposito) [reply app/scripts/metamask-controller.js]: Normalized in [01ca220](https://github.com/MetaMask/metamask-extension/pull/25435/commits/01ca22023cea120e40d81dbf5d00f34217be569d)

**Review by Gudahtt — DISMISSED** [review]
LGTM!

File: app/scripts/metamask-controller.js (hunk 5 above)
Comment (mikesposito) [reply app/scripts/metamask-controller.js]: I agree, `options` makes more sense! Changed in [ee14fba](https://github.com/MetaMask/metamask-extension/pull/25435/commits/ee14fbab78cd0e53f7c278ba4af2dae25f785444)

File: app/scripts/metamask-controller.js (hunk 4 above)
Comment (mikesposito) [reply app/scripts/metamask-controller.js:4325]: I agree. Also, ideally, side effects should be executed _after_ the account removal instead of _before_ since an error could occur (that now would also trigger a rollback)

**Review by Gudahtt — APPROVED** [review]

**Review by cryptodev-2s — APPROVED** [review]
LGTM!
